File: src/main/java/com/mercateo/jsonschema/generictype/GenericTypeHierarchy.java
package com.mercateo.jsonschema.generictype;

import javaslang.collection.Iterator;
import javaslang.collection.Stream;

public final class GenericTypeHierarchy {
    public Stream<GenericType<?>> hierarchy(GenericType<?> genericType) {
        return Stream.ofAll(new GenericTypeIterable(genericType));
    }

    private final static class GenericTypeIterable implements Iterable<GenericType<?>> {

        private final GenericType<?> genericType;

        GenericTypeIterable(GenericType<?> genericType) {
            this.genericType = genericType;
        }

        @Override
        public Iterator<GenericType<?>> iterator() {
            return new GenericTypeIterator(genericType);
        }
    }

    private final static class GenericTypeIterator implements Iterator<GenericType<?>> {
        GenericType<?> currentType;

        GenericTypeIterator(GenericType<?> genericType) {
            this.currentType = genericType;
        }

        @Override
        public boolean hasNext() {
            return currentType != null;
        }

        @Override
        public GenericType<?> next() {
            GenericType<?> type = currentType;
            currentType = currentType.getSuperType();
            return type;
        }
    }
}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericTypeHierarchy.kt
package com.mercateo.jsonschema.generictype


class GenericTypeHierarchy {
    fun hierarchy(genericType: GenericType<*>): Sequence<GenericType<*>> {

        return sequence {
            var currentType: GenericType<*>? = genericType

            while (currentType != null) {
                @Suppress("USELESS_CAST")
                yield(currentType as GenericType<*>)
                currentType = currentType.superType
            }
        }
    }
}
